src/rt-app_args.c

```diff
@@ -23,36 +23,79 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 #include <stdio.h>
 #include <string.h>
 #include <sys/stat.h>
+#include <getopt.h>
 
 #include "rt-app_parse_config.h"
 
+char help_usage[] = \
+"Usage: rt-app <taskset.json>\n"
+"Try 'rt-app --help' for more information.\n";
+
+char help_full[] = \
+"Usage:\n"
+"      rt-app <taskset.json>\n"
+"      cat taskset.json | rt-app -\n\n"
+"taskset.json is a json file describing the workload that will be generated"
+"by rt-app.\n\n"
+"In the first example, the json file is opened and parsed by rt-app.\n"
+"In the second example, rt-app reads the workload description in json format\n"
+"through the standard input.\n\n"
+"Miscellaneous:\n"
+"  -v, --version             display version information and exit\n"
+"  -h, --help                display this help text and exit\n";
+
 void
-usage (const char* msg, int ex_code)
+usage(const char* msg, int ex_code)
 {
-	printf("usage:\n"
-	       "rt-app <taskset.json>\n");
+	printf("%s", help_usage);
 
 	if (msg != NULL)
 		printf("\n%s\n", msg);
 	exit(ex_code);
 }
 
+struct option long_args[] = {
+	{"help",	no_argument,	0,	'h'},
+	{"version",	no_argument,	0,	'v'},
+	{0,		0,		0,	0}
+};
+
 void
 parse_command_line(int argc, char **argv, rtapp_options_t *opts)
 {
 	struct stat config_file_stat;
+	int c;
 
-	if (argc < 2)
-		usage(NULL, EXIT_SUCCESS);
+	while (1) {
+		c = getopt_long(argc, argv, "hv", long_args, 0);
+		if (c == -1)
+			break;
 
-	if (stat(argv[1], &config_file_stat) == 0) {
-		parse_config(argv[1], opts);
-		return;
-	} else if (strcmp(argv[1], "-") == 0) {
-		parse_config_stdin(opts);
-		return;
+		switch (c) {
+		case 'h':
+			printf("%s", help_full);
+			exit(0);
+			break;
+		case 'v':
+			printf("%s %s (%s)\n",
+			       PACKAGE,
+			       VERSION,
+			       BUILD_DATE);
+			exit(0);
+			break;
+		default:
+			usage(NULL, EXIT_INV_COMMANDLINE);
+			break;
+		}
 	}
 
-	usage(NULL, EXIT_SUCCESS);
-}
+	if (optind >= argc)
+		usage(NULL, EXIT_INV_COMMANDLINE);
 
+	if (stat(argv[optind], &config_file_stat) == 0)
+		parse_config(argv[optind], opts);
+	else if (strcmp(argv[optind], "-") == 0)
+		parse_config_stdin(opts);
+	else
+		usage(NULL, EXIT_FAILURE);
```

Comment:
Maybe the error it's more readable if you add an empty line before the usage notes.

Reply:
I agree, but this choice has been done to make it as similar as the popular tools:
```
$ tar
tar: You must specify one of the '-Acdtrux', '--delete' or '--test-label' options
Try 'tar --help' or 'tar --usage' for more information.

$ src/rt-app 
Usage: rt-app <taskset.json>
Try 'rt-app --help' for more information.
```
Does it make sense?

Reply:
ok, fine with me! ;-)